linked_list/implementation.py

```diff
@@ -1,40 +1,127 @@
+            element = self.start
+            while element.next:
+                yield element
+                element = element.next
+            # yield end element
+            yield element
+        
 
     def __getitem__(self, index):
-        pass
+        for i, element in enumerate(iter(self)):
+            if i == index:
+                return element
+        raise IndexError
 
     def __add__(self, other):
-        pass
+        new_list = LinkedList()
+        for e in iter(self):
+            new_list.append(e.elem)
+        for e in iter(other):
+            new_list.append(e.elem)
+        return new_list
 
     def __iadd__(self, other):
-        pass
+        if self.start == None:
+            self.start = other.start
+            self.end = other.end
+
+        elif other.start == None:
+            pass
+        else:
+            self.end.next = other.start
+            self.end = other.end
+        return self
+
 
     def __eq__(self, other):
-        pass
+        self_elem = self.start
+        other_elem = other.start
+        # check if list is empty or single item
+        if not self_elem or not self_elem.next:
+            return self_elem == other_elem
+        while self_elem.next:
+            if self_elem.elem != other_elem.elem:
+                return False
+            self_elem = self_elem.next
+            other_elem = other_elem.next
+        return True
+
+    def __ne__(self, other):
+        self_elem = self.start
+        other_elem = other.start
+        # check if list is empty or single item
+        if not self_elem or not self_elem.next:
+            return self_elem != other_elem
+        while self_elem.next:
+            if self_elem != other_elem:
+                return True
+            self_elem = self_elem.next
+            other_elem = other_elem.next
+        return False
+            
 
     def append(self, elem):
-        pass
+        new_end = Node(elem)
+        if not self.start:
+            self.start = new_end
+            self.end = new_end
+        else:
+            self.end.next = new_end
+            self.end = new_end
 
     def count(self):
-        pass
+        return len(self)
 
     def pop(self, index=None):
-        pass
+        if index is None:
+            index = self.count() -1
+        previous_element = None
+        for i, element in enumerate(iter(self)):
+            if i == index:
+                if element == self.start and element == self.end:
+                    self.start = None
+                    self.end = None
+                elif element == self.start:
+                    self.start = self.start.next
+                elif element == self.end:
+                    self.end == previous_element
+                    previous_element.next = None
+                else:
+                    previous_element.next = element.next
+                return element
+            previous_element = element
+        raise IndexError
```

Comment:
Wow that's a really clean and clever code. I like it. I hade to raise `IndexError` at 2 different lines to make my code pass the tests. Thumbs up!